### PR TITLE
Hook for BaseScope.Apply

### PR DESCRIPTION
**Pull request by Tyrrrz**
Related to #594

@bruno-garcia I've looked at how existing `Func<IEnumerable<ISentryEventProcessor>>[]` and etc. work, but they don't make sense to me. Can you guide me in regards to what I should do to make it usable from ASP.NET Core?

### CODE REVIEW

**Comment by Tyrrrz** [comment]
Open question: what should happen in the event of `scope.IsLocked`? Should the apply be called on the current scope? Should it only happen if the custom hook was provided?

**Comment by bruno-garcia** [comment]
> Open question: what should happen in the event of scope.IsLocked? Should the apply be called on the current scope? Should it only happen if the custom hook was provided?

yeah we just set the data to the current scope.

File: src/Sentry.Protocol/BaseScopeExtensions.cs [hunk 1]
```diff
@@ -308,38 +307,8 @@ public static void Apply(this BaseScope from, BaseScope to)
...
-                    {
-                        scope.SetTag(tupleStringString.Item1, tupleStringString.Item2);
-                    }
-                    break;
-#endif
-                default:
-                    scope.SetExtra("state", state);
-                    break;
-            }
+            var processor = scope.ScopeOptions.SentryScopeProcessor ?? new DefaultSentryScopeProcessor();
```
Comment (bruno-garcia) [inline src/Sentry.Protocol/BaseScopeExtensions.cs:310]: We'll allocate this processor for each call to `Apply`, what about we set this instance to the Options ctor (and make it non null), users can set a new one and replace it though

**Review by bruno-garcia — COMMENTED** [review]
Looking good!

Given that the high level API is `X<TScope>(T arg` I wonder if we could do the same here and dump the `object` arg

File: src/Sentry.Protocol/BaseScopeExtensions.cs (hunk 1 above)
Comment (Tyrrrz) [reply src/Sentry.Protocol/BaseScopeExtensions.cs:310]: It is already like that on `SentryOptions`, but we can't guarantee it on the interface level (`IScopeOptions`) unless we unconditionally trust that user will respect NRTs. The RHS of `??` should not be executed majority of the time.

**Pull request merged**